patches/src/main/kotlin/app/revanced/patches/shared/misc/mapping/ResourceMappingPatch.kt

```diff
@@ -1,64 +1,70 @@
 import app.revanced.patcher.patch.resourcePatch
 import org.w3c.dom.Element
-import java.util.*
-import java.util.concurrent.Executors
-import java.util.concurrent.TimeUnit
+import java.util.Collections
 
-// TODO: Probably renaming the patch/this is a good idea.
-lateinit var resourceMappings: List<ResourceElement>
-    private set
+data class ResourceElement(val type: String, val name: String, val id: Long)
 
-val resourceMappingPatch = resourcePatch {
-    val resourceMappings = Collections.synchronizedList(mutableListOf<ResourceElement>())
-
-    execute {
-        val threadCount = Runtime.getRuntime().availableProcessors()
-        val threadPoolExecutor = Executors.newFixedThreadPool(threadCount)
+private lateinit var resourceMappings: MutableMap<String, ResourceElement>
 
-        // Save the file in memory to concurrently read from it.
-        val resourceXmlFile = get("res/values/public.xml").readBytes()
-
-        for (threadIndex in 0 until threadCount) {
-            threadPoolExecutor.execute thread@{
-                document(resourceXmlFile.inputStream()).use { document ->
+private fun setResourceId(type: String, name: String, id: Long) {
+    resourceMappings[type + name] = ResourceElement(type, name, id)
+}
 
-                    val resources = document.documentElement.childNodes
-                    val resourcesLength = resources.length
-                    val jobSize = resourcesLength / threadCount
+/**
+ * @return A resource id of the given resource type and name.
+ * @throws PatchException if the resource is not found.
+ */
+fun getResourceId(type: String, name: String) = resourceMappings[type + name]?.id
+    ?: throw PatchException("Could not find resource type: $type name: $name")
+
+/**
+ * @return All resource elements.  If a single resource id is needed instead use [getResourceId].
+ */
+fun getResourceElements() = Collections.unmodifiableCollection(resourceMappings.values)
+
+/**
+ * @return If the resource exists.
+ */
+fun hasResourceId(type: String, name: String) = resourceMappings[type + name] != null
+
+/**
+ * Identical to [LiteralFilter] except uses a decoded resource literal value.
+ *
+ * Any patch with fingerprints of this filter must
+ * also declare [resourceMappingPatch] as a dependency.
+ */
+fun resourceLiteral(
+    type: String,
+    name: String,
+    maxBefore: Int = METHOD_MAX_INSTRUCTIONS,
+) = literal({ getResourceId(type, name) }, null, maxBefore)
 
-                    val batchStart = jobSize * threadIndex
-                    val batchEnd = jobSize * (threadIndex + 1)
-                    element@ for (i in batchStart until batchEnd) {
-                        // Prevent out of bounds.
-                        if (i >= resourcesLength) return@thread
 
-                        val node = resources.item(i)
-                        if (node !is Element) continue
+val resourceMappingPatch = resourcePatch {
+    execute {
+        document("res/values/public.xml").use { document ->
```

Comment:
```suggestion
        document(get("res/values/public.xml").inputStream()).use { document ->
```

Previously, I said "Because the concurrent execution was removed, this can be simplified
`document("res/values/public.xml")`" at https://github.com/ReVanced/revanced-patches/pull/4409#discussion_r1964974416 , but this causes unnecessary file overwriting.
You had to use inputStream (read-only) version of `Document` constructor.
Sorry for my mistake.

But `.readBytes()` is still unnecessary since concurrent execution was removed.

Also, you can remove `.use {}` block because it is read-only version of `Document`. (This is not very important though)